### ISSUE TITLE
[MIRA] Refining the model using Bayesian optimization

### ISSUE THREAD

**Issue by jpark27** [cistrome/MIRA#53]
Dear MIRA team,

Hi, all! Thanks so much for sharing great tool with the community.

I have multiome data with 68 K cells and 110K ATAC peaks.

For RNA compartment, following tutorial, I can decide topic number (= 18) with gradient tuning as follows. example_rna_model.trim_learning_rate_bounds(2.5, 1)

<img width="2825" height="960" alt="Image" src="https://github.com/user-attachments/assets/5db1701b-50e0-4a65-aeb2-0d367fff24c3" />

<img width="2224" height="960" alt="Image" src="https://github.com/user-attachments/assets/3cb10ceb-6dcd-4db1-ad5c-e186824764c1" />

Then, ATAC compartment, following another tutorial (https://mira-multiome.readthedocs.io/en/latest/notebooks/tutorial_atlas_integration.html), gradient tuning suggest the topic number is only (=1). I set the boundary, atac_model.set_learning_rates(10**(-3), 10**(-0.1)).

<img width="4111" height="1915" alt="Image" src="https://github.com/user-attachments/assets/72853224-5866-4a0a-92c2-9513c7d301d5" />

<img width="2712" height="960" alt="Image" src="https://github.com/user-attachments/assets/9c2c3023-f1bb-4427-b71b-92496f391323" />

Hence, I decided to continue with the Bayesian optimization. 

tuner = mira.topics.BayesianTuner(
    model = atac_model,
    min_topics = NUM_TOPICS,
    max_topics = NUM_TOPICS + 19,
    save_name = '/project/simmons_hts/jpark/0_tools/HALO/MIRA/IBD_CD4/2',
    n_jobs=4
    # n_jobs = 5, # highly suggested - if you have the GPU memory!
    # storage = mira.topics.Redis(), # if you are running a redis server backend
)

tuner.fit('/project/simmons_hts/jpark/0_tools/HALO/MIRA/atac_train_2', '/project/simmons_hts/jpark/0_tools/HALO/MIRA/atac_test_2')

With GPU, overnight run, most of runs throw ERROR like following.

###############################################################################################

Trials finished: 80 | Best trial: 1 | Best score: -2.9082e+04
Press ctrl+C,ctrl+C or esc,I+I,I+I in Jupyter notebook to stop early.

Tensorboard logidr: /project/simmons_hts/jpark/0_tools/HALO/MIRA/IBD_CD4/2
#Topics | #Trials

      1 | ■ ■ ■
      2 | ■ ■ ■
      3 | ■ ■ ■ ■ ■
      4 | ■ ■ ■
      5 | ■
      6 | ■ ■ ■ ■ ■ ■
      7 | ■
      8 | ■ ■
      9 | ■ ■ ■
     10 | ■ ■ ■ ■
     11 | ■ ■ ■
     12 | ■ ■ ■ ■ ■ ■
     13 | ■ ■
     14 | ■ ■
     15 | ■ ■
     16 | ■ ■ ■ ■
     17 | ■ ■ ■ ■ ■
     18 | ■ ■ ■ ■ ■ ■ ■ ■
     19 | ■ ■ ■ ■
     20 | ■ ■ ■ ■ ■ ■

Trial | Result (● = best so far)         | Params
 #0   |   | ERROR                        | {'decoder_dropout': 0.060706686870759155, 'num_topics': 12}
 #1   | ● | completed, score: -2.9082e+04 | {'decoder_dropout': 0.0626, 'num_topics': 2}
 #2   |   | ERROR                        | {'decoder_dropout': 0.0643607214055538, 'num_topics': 12}
 #3   |   | ERROR                        | {'decoder_dropout': 0.061449853687583766, 'num_topics': 7}
 #4   |   | ERROR                        | {'decoder_dropout': 0.06377579078089535, 'num_topics': 3}
 #5   |   | ERROR                        | {'decoder_dropout': 0.050809722605645175, 'num_topics': 4}
 #6   |   | ERROR                        | {'decoder_dropout': 0.05560198201322331, 'num_topics': 18}
 #7   |   | ERROR                        | {'decoder_dropout': 0.057593882672002474, 'num_topics': 6}
 #8   |   | ERROR                        | {'decoder_dropout': 0.05402416848275977, 'num_topics': 12}
 #9   |   | ERROR                        | {'decoder_dropout': 0.050942250569730355, 'num_topics': 8}
 #10  |   | ERROR                        | {'decoder_dropout': 0.060764308992040356, 'num_topics': 18}
 #11  |   | ERROR                        | {'decoder_dropout': 0.06462469771588296, 'num_topics': 3}
 #12  |   | ERROR                        | {'decoder_dropout': 0.05162277433299546, 'num_topics': 20}
 #13  |   | ERROR                        | {'decoder_dropout': 0.05064465235796102, 'num_topics': 4}
 #14  |   | ERROR                        | {'decoder_dropout': 0.051396322049259896, 'num_topics': 6}
 #15  |   | ERROR                        | {'decoder_dropout': 0.05460580073570787, 'num_topics': 17}
 #16  |   | ERROR                        | {'decoder_dropout': 0.05450977364835228, 'num_topics': 1}
 #17  |   | ERROR                        | {'decoder_dropout': 0.05650169388754086, 'num_topics': 18}
 #18  |   | ERROR                        | {'decoder_dropout': 0.05580243082567579, 'num_topics': 18}
 #19  |   | ERROR                        | {'decoder_dropout': 0.06054978207758301, 'num_topics': 11}
 #20  |   | ERROR                        | {'decoder_dropout': 0.051079831676910666, 'num_topics': 20}
 #21  |   | ERROR                        | {'decoder_dropout': 0.05205033826064811, 'num_topics': 10}
 #22  |   | ERROR                        | {'decoder_dropout': 0.056949761022360226, 'num_topics': 16}
 #23  |   | ERROR                        | {'decoder_dropout': 0.06107578478165173, 'num_topics': 11}
 #24  |   | ERROR                        | {'decoder_dropout': 0.06232139945508288, 'num_topics': 18}
 #25  |   | pruned at step: 1            | {'decoder_dropout': 0.0623, 'num_topics': 3}
 #26  |   | ERROR                        | {'decoder_dropout': 0.05343873756215455, 'num_topics': 2}
 #27  |   | ERROR                        | {'decoder_dropout': 0.05422888138638984, 'num_topics': 11}
 #28  |   | pruned at step: 1            | {'decoder_dropout': 0.0622, 'num_topics': 4}
 #29  |   | ERROR                        | {'decoder_dropout': 0.05360644782777066, 'num_topics': 16}
 #30  |   | ERROR                        | {'decoder_dropout': 0.062393036114156857, 'num_topics': 9}
 #31  |   | ERROR                        | {'decoder_dropout': 0.06367506127603575, 'num_topics': 17}
 #32  |   | ERROR                        | {'decoder_dropout': 0.05246775323773145, 'num_topics': 10}
 #33  |   | ERROR                        | {'decoder_dropout': 0.05232256641297127, 'num_topics': 13}
 #34  |   | ERROR                        | {'decoder_dropout': 0.05612287174238694, 'num_topics': 17}
 #35  |   | ERROR                        | {'decoder_dropout': 0.05902539458103327, 'num_topics': 9}
 #36  |   | ERROR                        | {'decoder_dropout': 0.06095560695507456, 'num_topics': 19}
 #37  |   | ERROR                        | {'decoder_dropout': 0.05494485232542869, 'num_topics': 6}
 #38  |   | pruned at step: 1            | {'decoder_dropout': 0.0545, 'num_topics': 3}
 #39  |   | ERROR                        | {'decoder_dropout': 0.05007590477027371, 'num_topics': 18}
 #40  |   | ERROR                        | {'decoder_dropout': 0.0648160993611371, 'num_topics': 1}
 #41  |   | ERROR                        | {'decoder_dropout': 0.06140034728136951, 'num_topics': 6}
 #42  |   | ERROR                        | {'decoder_dropout': 0.057221662949952824, 'num_topics': 14}
 #43  |   | pruned at step: 8            | {'decoder_dropout': 0.0563, 'num_topics': 5}
 #44  |   | ERROR                        | {'decoder_dropout': 0.05116868151940848, 'num_topics': 20}
 #45  |   | ERROR                        | {'decoder_dropout': 0.050043521631943755, 'num_topics': 3}
 #46  |   | ERROR                        | {'decoder_dropout': 0.06396156786714743, 'num_topics': 17}
 #47  |   | ERROR                        | {'decoder_dropout': 0.05697679066622769, 'num_topics': 1}
 #48  |   | ERROR                        | {'decoder_dropout': 0.057619426533495326, 'num_topics': 14}
 #49  |   | ERROR                        | {'decoder_dropout': 0.06053367034833039, 'num_topics': 2}
 #50  |   | ERROR                        | {'decoder_dropout': 0.05036920870872102, 'num_topics': 15}
 #51  |   | ERROR                        | {'decoder_dropout': 0.05551762747226601, 'num_topics': 19}
 #52  |   | ERROR                        | {'decoder_dropout': 0.06177905596361966, 'num_topics': 12}
 #53  |   | ERROR                        | {'decoder_dropout': 0.05085830753179675, 'num_topics': 18}
 #54  |   | ERROR                        | {'decoder_dropout': 0.06248058985388094, 'num_topics': 15}
 #55  |   | ERROR                        | {'decoder_dropout': 0.06489575882925118, 'num_topics': 20}
 #56  |   | ERROR                        | {'decoder_dropout': 0.05604642214343695, 'num_topics': 20}
 #60  |   | ERROR                        | {'decoder_dropout': 0.05843449090162245, 'num_topics': 20}
 #62  |   | ERROR                        | {'decoder_dropout': 0.055867311806180396, 'num_topics': 13}
 #63  |   | ERROR                        | {'decoder_dropout': 0.05371071704885484, 'num_topics': 12}
 #64  |   | ERROR                        | {'decoder_dropout': 0.052516905320548125, 'num_topics': 17}
 #65  |   | ERROR                        | {'decoder_dropout': 0.05276071717099527, 'num_topics': 9}
 #66  |   | ERROR                        | {'decoder_dropout': 0.057544329922898305, 'num_topics': 19}
 #67  |   | ERROR                        | {'decoder_dropout': 0.05806324696201364, 'num_topics': 6}
 #68  |   | ERROR                        | {'decoder_dropout': 0.05540311724373809, 'num_topics': 6}
 #69  |   | ERROR                        | {'decoder_dropout': 0.05671935888262574, 'num_topics': 16}
 #70  |   | ERROR                        | {'decoder_dropout': 0.05179665465648481, 'num_topics': 10}
 #71  |   | ERROR                        | {'decoder_dropout': 0.054213439280428447, 'num_topics': 19}
 #72  |   | ERROR                        | {'decoder_dropout': 0.05221219101312983, 'num_topics': 12}
 #73  |   | ERROR                        | {'decoder_dropout': 0.05127656990742778, 'num_topics': 18}
 #74  |   | ERROR                        | {'decoder_dropout': 0.051839591356356185, 'num_topics': 16}
 #75  |   | ERROR                        | {'decoder_dropout': 0.05526610614956157, 'num_topics': 10}
 #77  |   | ERROR                        | {'decoder_dropout': 0.060423623509834996, 'num_topics': 8}

Running trials:
Trial | Progress                         | Params
 #57  |                                  | {'decoder_dropout': 0.0577, 'num_topics': 2}
 #58  |                                  | {'decoder_dropout': 0.0545, 'num_topics': 16}
 #59  |                                  | {'decoder_dropout': 0.0559, 'num_topics': 5}
 #61  |                                  | {'decoder_dropout': 0.0599, 'num_topics': 18}
 #76  |                                  | {'decoder_dropout': 0.0574, 'num_topics': 15}
 #78  |                                  | {'decoder_dropout': 0.0593, 'num_topics': 3}
 #79  |                                  | {'decoder_dropout': 0.0586, 'num_topics': 8}
 #80  |                                  | {'decoder_dropout': 0.0586, 'num_topics': 18}

###############################################################################################

**1. Is there anyway to avoid error during bayesian tuning?
2. If it's unavoidable, on my current dataset, can I roughly go ATAC topic as (=18) just like RNA topic?** I recall there was a note that ATAC topic would be +-3 that of #RNA topics.

Any feedback would be really appreciated,
Jun
